src/controller/CHIPCommissionableNode.cpp

```diff
@@ -0,0 +1,151 @@
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      Implementation of CHIP Device Controller, a common class
+ *      that implements discovery, pairing and provisioning of CHIP
+ *      devices.
+ *
+ */
+
+#ifndef __STDC_LIMIT_MACROS
+#define __STDC_LIMIT_MACROS
+#endif
+#ifndef __STDC_FORMAT_MACROS
+#define __STDC_FORMAT_MACROS
+#endif
+
+// module header, comes first
+#include <controller/CHIPCommissionableNode.h>
+
+#if CONFIG_DEVICE_LAYER
+#include <platform/CHIPDeviceLayer.h>
+#endif
+
+#include <core/CHIPCore.h>
+#include <core/CHIPEncoding.h>
+#include <core/CHIPSafeCasts.h>
+#include <support/CHIPArgParser.hpp>
+#include <support/CHIPMem.h>
+#include <support/CodeUtils.h>
+#include <support/logging/CHIPLogging.h>
+
+#if CONFIG_NETWORK_LAYER_BLE
+#include <ble/BleLayer.h>
+#include <transport/raw/BLE.h>
+#endif
+
+#include <app/util/af-enums.h>
+
+#include <errno.h>
+#include <inttypes.h>
+#include <memory>
+#include <stdint.h>
+#include <stdlib.h>
+
+namespace chip {
+namespace Controller {
+
+CHIP_ERROR CommissionableNode::Init()
+{
+    if (Platform::MemoryInit() != CHIP_NO_ERROR)
+    {
+        ChipLogError(chipTool, "FAILED to initialize memory");
+        return 1;
+    }
+
+#if CONFIG_DEVICE_LAYER
+    if (DeviceLayer::PlatformMgr().InitChipStack() != CHIP_NO_ERROR)
+    {
+        ChipLogError(chipTool, "FAILED to initialize chip stack");
+        return 1;
+    }
+#endif
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CommissionableNode::DiscoverAllCommissioners()
+{
+    chip::Mdns::Resolver::Instance().SetResolverDelegate(this);
+#if CONFIG_DEVICE_LAYER
+    ReturnErrorOnFailure(chip::Mdns::Resolver::Instance().StartResolver(&DeviceLayer::InetLayer, kMdnsPort));
+#endif
+
+    for (int i = 0; i < kMaxCommissioners; ++i)
+    {
+        mCommissioners[i].Reset();
+    }
+    return chip::Mdns::Resolver::Instance().FindCommissioners();
+}
+
+CHIP_ERROR CommissionableNode::DiscoverAllCommissionersLongDiscriminator(uint16_t long_discriminator)
+{
+    chip::Mdns::Resolver::Instance().SetResolverDelegate(this);
+#if CONFIG_DEVICE_LAYER
+    ReturnErrorOnFailure(chip::Mdns::Resolver::Instance().StartResolver(&DeviceLayer::InetLayer, kMdnsPort));
+#endif
+
+    for (int i = 0; i < kMaxCommissioners; ++i)
+    {
+        mCommissioners[i].Reset();
+    }
+    Mdns::DiscoveryFilter filter(Mdns::DiscoveryFilterType::kLong, long_discriminator);
+    return Mdns::Resolver::Instance().FindCommissionableNodes(filter);
+}
+
+const Mdns::CommissionableNodeData * CommissionableNode::GetDiscoveredDevice(int idx)
+{
+    // TODO(cecille): Add assertion about main loop.
```

Comment:
is this copy&pasted code? Could we centralize code instead of just copying?

Reply:
I had to refactor some of the pre-existing code to avoid code duplication and found that the change touched a lot of files. In order to not confuse the new logic in this PR focussed on introducing Commissioner Discovery with the larger refactor, I have posted the refactor as a follow-up PR: https://github.com/sharadb-amazon/connectedhomeip/pull/1 

@cecille  / @andy31415 / @chrisdecenzo  - Do you mind taking a look at the refactor PR: https://github.com/sharadb-amazon/connectedhomeip/pull/1 ?